src/hotspot/share/nmt/memMapPrinter.cpp

```diff
@@ -0,0 +1,291 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "logging/logAsyncWriter.hpp"
+#include "gc/shared/collectedHeap.hpp"
+#include "memory/allocation.hpp"
+#include "memory/universe.hpp"
+#include "runtime/nonJavaThread.hpp"
+#include "runtime/osThread.hpp"
+#include "runtime/thread.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/vmThread.hpp"
+#include "nmt/memFlagBitmap.hpp"
+#include "nmt/memMapPrinter.hpp"
+#include "nmt/memTracker.hpp"
+#include "nmt/virtualMemoryTracker.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/ostream.hpp"
+
+// Note: throughout this code we will use the term "VMA" for OS system level memory mapping
+
+/// NMT mechanics
+
+// Short, clear, descriptive names for all possible markers. Note that we only expect to see
+// those that have been used with mmap. Flags left out are printed with their nmt flag name.
+#define NMT_FLAGS_DO(f) \
+  /* flag, short, description */ \
+  f(mtGCCardSet,      "CARDTBL", "GC Card table") \
+  f(mtClassShared,    "CDS", "CDS archives") \
+  f(mtClass,          "CLASS", "Class Space") \
+  f(mtCode,           "CODE", "Code Heap") \
+  f(mtGC,             "GC", "GC support data (e.g. bitmaps)") \
+  f(mtInternal,       "INTERN", "Internal") \
+  f(mtJavaHeap,       "JAVAHEAP", "Java Heap") \
+  f(mtOther,          "JDK", "allocated by JDK libraries other than VM") \
+  f(mtMetaspace,      "META", "Metaspace nodes (non-class)") \
+  f(mtSafepoint,      "POLL", "Polling pages") \
+  f(mtThreadStack,    "STACK", "(known) Thread Stack") \
+  f(mtTest,           "TEST", "JVM internal test mappings")
+  //end
+
+static const char* get_shortname_for_nmt_flag(MEMFLAGS f) {
+#define DO(flag, shortname, text) if (flag == f) return shortname;
+  NMT_FLAGS_DO(DO)
+#undef DO
+  return NMTUtil::flag_to_enum_name(f);
+}
+
+/// NMT virtual memory
+
+static bool range_intersects(const void* from1, const void* to1, const void* from2, const void* to2) {
+  const void* const min = MAX2(from1, from2);
+  const void* const max = MIN2(to1, to2);
+  return min < max;
+}
+
+// A Cache that correlates range with MEMFLAG, optimized to be iterated quickly
+// (cache friendly).
+class CachedNMTInformation : public VirtualMemoryWalker {
+  struct Range { const void* from; const void* to; };
+  // Unfortunately, we need to allocate manually, raw, since we must prevent
+  // NMT deadlocks (ThreadCritical).
+  Range* _ranges;
+  MEMFLAGS* _flags;
+  uintx _count, _capacity;
+public:
+  CachedNMTInformation() : _ranges(nullptr), _flags(nullptr), _count(0), _capacity(0) {}
+
+  ~CachedNMTInformation() {
+    ALLOW_C_FUNCTION(free, ::free(_ranges);)
+    ALLOW_C_FUNCTION(free, ::free(_flags);)
+  }
+
+  bool add(const void* from, const void* to, MEMFLAGS f) {
```

Comment:
Please mention that we're `add`ing in sorted order, that is that `forall R \in _ranges R.to <= from` holds.

Reply:
I wasn't sure about that. Do we always? Are NMT regions guaranteed to be sorted? (Dimly remember cases where that weren't so)

Reply:
It's always sorted in some way, we save the linked lists in some different sorting orders for baselining, but in the `VirtualMemoryTracker` they're guaranteed to be sorted by address.